Update README for GA release (#205)
cyberark/conjur-api-python3: README.md

# conjur-api-python3

> Command-line tool and programmatic Python access to the Conjur API to manage your Conjur resources. 

Reviewer: Sigal I think we need to seperate the 2 offering this repo has:

1. Command-line tool to manage your Conjur resources. This CLI is self-contained.
2. Python SDK based on Conjur API, written in Python3

combining the 2 just makes it harder to read

Reviewer 2: Is this in addition to the above sentence that yoou wat to add @sigalsax ??

From @InbalZilberman's comment:
- A command-line tool for managing your Conjur resources. This Conjur CLI is self-contained.
- Python SDK based on the Conjur REST API, written in Python3

Author: The follow-up sentence explains it a bit more

"This repository includes both the self-contained Conjur command-line tool (`conjur`) and Python3-based SDK for Conjur."

I prefer we leave it because both are used to manage Conjur resources. In the follow-up sentence we explain a bit more

Reviewer: Sure I just think it is best to seperate the 2 as much as possible in the text and in visualization this is why i suggest: 1,2

> This repository includes both the self-contained Conjur command-line tool (`conjur`) and Python3-based SDK for Conjur.

Reviewer: I like this sentence much more :) but it feels a bit of repetition of the first one, dont you think?

Author: what about "This repository includes both the self-contained Conjur command-line tool (`conjur`) and Python3-based SDK for access to the Conjur API to manage your Conjur resources."

[![Test Coverage](https://api.codeclimate.com/v1/badges/d90d69a3942120b36785/test_coverage)](https://codeclimate.com/github/cyberark/conjur-api-python3/test_coverage) [![Maintainability](https://api.codeclimate.com/v1/badges/d90d69a3942120b36785/maintainability)](https://codeclimate.com/github/cyberark/conjur-api-python3/maintainability)

---

## Certificate level

### CLI

![](https://img.shields.io/badge/Certification%20Level-Certified-6C757D?link=https://github.com/cyberark/community/blob/master/Conjur/conventions/certification-levels.md)

The Conjur CLI is a **Certified** level project. It's been reviewed by CyberArk to verify that it will securely
work with CyberArk DAP as documented. In addition, CyberArk offers Enterprise-level support for these features. For
more detailed information on our certification levels, see [our community guidelines](https://github.com/cyberark/community/blob/master/Conjur/conventions/certification-levels.md#community).

### SDK

![](https://img.shields.io/badge/Certification%20Level-Community-28A745?link=https://github.com/cyberark/community/blob/master/Conjur/conventions/certification-levels.md)

> The Conjur Python SDK is a **Community** level project. It's a community contributed project that **is not reviewed or supported

Reviewer 2: ```suggestion
The Conjur Python SDK is a **Community** level project. It's a community contributed projec, and  **is not reviewed or supported
```

Author: @izgeri might need to update the general description on your end as well?

> by CyberArk**. For more detailed information on our certification levels, see [our community guidelines](https://github.com/cyberark/community/blob/master/Conjur/conventions/certification-levels.md#community).

Reviewer 2: below (line 32): Conjur OSS Suite (capital  S)
and line 33... בקיצור everwhere... Suite is with a capital S

(line 34) updgrade your Conjur version -> your Conjur OSS Suite version

Reviewer 2: "when using integrations, choose the..."
change to "When (capital W). Full stop before it instead of semi-colon.

Reviewer 2: choose the latest suite release that matches your Conjur **OSS** version.

# conjur-api-python3

> Python3-based API SDK for [Conjur OSS](https://www.conjur.org/). The repo

Author: Conjur CLI / SDK lets you access and manage Conjur resources from either the command line or from your application

Reviewer 2: Using the Conjur CLI / SD,  you can access and manage Conjur resources from the command line or from your application

> also includes a self-contained CLI tool (`conjur-cli`) that wraps the API

Author: Conjur instead of cli

> The repo also includes a self-contained CLI tool (`conjur-cli`) that wraps the API

Author: conjur instead of conjur-cli

> in a simple executable script/binary.

Author: @izgeri do you know what does this mean?

Reviewer 3: can you say which part is not clear? it's hard to tell what you're referring to from this comment

Author: "that wraps the API in a simple executable script/binary."
The executable script/binary part is not clear. Like the client wraps the API but as a library that can be imported so I don't understand what it means "wraps the API in a simple executable script/binary"

Reviewer 3: I believe it's saying that the CLI is a wrapper around the Python client, and the CLI is an executable script/binary

[![Test Coverage](https://api.codeclimate.com/v1/badges/d90d69a3942120b36785/test_coverage)](https://codeclimate.com/github/cyberark/conjur-api-python3/test_coverage) [![Maintainability](https://api.codeclimate.com/v1/badges/d90d69a3942120b36785/maintainability)](https://codeclimate.com/github/cyberark/conjur-api-python3/maintainability)

---

> ### **Status**: GA

Reviewer: Is GA is for SDK and CLI?
We need to make it GA only once we release it... after signing & testing...

Author: This is a good question. What requirements are need to be met in order for client as well to be GA level?

Reviewer: 1. product and UX review
2. documentation
3. test coverage
4. security review 

Basically our regular DOD
Also the link it self can help:
https://github.com/cyberark/community/blob/master/Conjur/conventions/certification-levels.md

In any case CLI is planned for GA :)

Author: so our plan is to take the CLI to GA but we don't have different versions for the CLI and SDK in PyPi because they are shipped together so I think
How do we usually handle this? @InbalZilberman @izgeri ? Can we promote one but not the other? How would versioning work?

> ### **Status**: GA

Reviewer 3: We should actually use the OSS standard for this section in the README with the GA release - does this project meet the standards for Certified? Is it going to be published in the documentation? Are the CLI AND the client library certified, or just the CLI?
```suggestion
## Certification level

{Certified}
![](https://img.shields.io/badge/Certification%20Level-Certified-6C757D?link=https://github.com/cyberark/community/blob/master/Conjur/conventions/certification-levels.md)

This repo is a **Certified** level project. It's been reviewed by CyberArk to verify that it will securely
work with CyberArk DAP as documented. In addition, CyberArk offers Enterprise-level support for these features. For
more detailed  information on our certification levels, see [our community guidelines](https://github.com/cyberark/community/blob/master/Conjur/conventions/certification-levels.md#community).

```

Author: @izgeri would you give the SDK a community level or trusted level badge? 
Because for community the description is "It's a community contributed project that is not reviewed or supported by CyberArk." which is not exactly true

# conjur-api-python3

Command-line tool and programmatic Python access to the Conjur API to manage your Conjur resources. 

This repository includes both the self-contained Conjur command-line tool (`conjur`) and Python3-based SDK for Conjur.

[![Test Coverage](https://api.codeclimate.com/v1/badges/d90d69a3942120b36785/test_coverage)](https://codeclimate.com/github/cyberark/conjur-api-python3/test_coverage) [![Maintainability](https://api.codeclimate.com/v1/badges/d90d69a3942120b36785/maintainability)](https://codeclimate.com/github/cyberark/conjur-api-python3/maintainability)

---

## Certificate level

### Conjur CLI

![](https://img.shields.io/badge/Certification%20Level-Certified-6C757D?link=https://github.com/cyberark/community/blob/master/Conjur/conventions/certification-levels.md)

The Conjur CLI is a **Certified** level project. It's been reviewed by CyberArk to verify that it will securely

> work with CyberArk DAP as documented. In addition, CyberArk offers Enterprise-level support for these features. For

Reviewer 2: not sure what you want to do about DAP vs CE at this point

Author: @InbalZilberman ?

Author: It will be a separate effort I believe. A separate effort dedicated to this name change

The Conjur Python SDK is a **Community** level project. It's a community contributed project that **is not reviewed or supported
by CyberArk**. For more detailed information on our certification levels, see [our community guidelines](https://github.com/cyberark/community/blob/master/Conjur/conventions/certification-levels.md#community).

### Using conjur-api-python3 with Conjur OSS 

Are you using this project with [Conjur OSS](https://github.com/cyberark/conjur)? Then we 
**strongly** recommend choosing the version of this project to use from the latest [Conjur OSS 
suite release](https://docs.conjur.org/Latest/en/Content/Overview/Conjur-OSS-Suite-Overview.html). 
Conjur maintainers perform additional testing on the suite release versions to ensure 
compatibility. When possible, upgrade your Conjur OSS version to match the 
[latest suite release](https://docs.conjur.org/Latest/en/Content/ReleaseNotes/ConjurOSS-suite-RN.htm).
When using integrations, choose the latest suite release that matches your Conjur OSS version. For any 
questions, please contact us on [Discourse](https://discuss.cyberarkcommons.org/c/conjur/5).

### Supported Services

> - Conjur OSS v1.2.0 or later

Reviewer 2: @sigalsax  Didn't we write in the docs v1.5.0 or later?

Author: should be 1.2

- Conjur Enterprise v11.2.1 (v5.6.3) or later

### Sypported Platforms

- macOS Catalina or later
- Windows 10 or later
- Red Hat Enterprise Linux 7, 8

### Installation

#### Install the Conjur CLI

To access the latest release of the Conjur CLI, go to our [release](https://github.com/cyberark/conjur-api-python3/releases) 

> page. For instructions on how to setup and configure the CLI, see our [official documentation](https://docs.conjur.org/Latest/en/Content/Developer/CLI/cli-lp.htm).

Reviewer 2: ```suggestion
page. For instructions on how to set up and configure the CLI, see our [official documentation](https://docs.conjur.org/Latest/en/Content/Developer/CLI/cli-lp.htm).
```

#### Install the SDK

The SDK can be installed via PyPI. Note that the SDK is a **Community** level project meaning 
that the SDK is subject to alterations that may result in breaking change. 

To avoid unanticipated breaking changes, make sure that you stay up-to-date on our 
latest releases and review the project's [CHANGELOG.md](CHANGELOG.md).

```
$ pip3 install conjur-client

$ conjur --help
```

Alternatively, you can install the library from the source. Note that this will install the latest work from the
cloned source and not necessarily an official release. 

Clone the project and run:

```
$ pip3 install .
```

## Usage

### CLI

> For more information on how to setup, configure, and start using the Conjur CLI, see our [official 

Reviewer 2: ```suggestion
For more information on how to set up, configure, and start using the Conjur CLI, see our [official 
```

when using integrations, choose the latest suite release that matches your Conjur version. For any 
questions, please contact us on [Discourse](https://discuss.cyberarkcommons.org/c/conjur/5).

### Support Services

- Conjur ***TODO find lowest version***
- DAP ***TODO find lowest version***

### Installation

#### Install the CLI

You can access the latest release of the CLI to install on your machine by navigating to our 
[release](https://github.com/cyberark/conjur-api-python3/releases) page. For instructions on how to setup and configure the CLI, see our 
official documentation ***(TODO add link to docs)*** 

> If Python is installed on your machine, you can install the client library by pulling from PyPI:

Reviewer: @sigalsax shouldn't we link to our docs also here?

Author: don't think so because nothing more is needed beyond pulling from pypi

when using integrations, choose the latest suite release that matches your Conjur version. For any 
questions, please contact us on [Discourse](https://discuss.cyberarkcommons.org/c/conjur/5).

### Supported Services

> - Conjur OSS v1.2.0+

Reviewer 2: @sigalsax , check the Conjur OSS readmes for the exact suite version numbers
I believe the one we discussed is v1.5.0+suite1

https://docs.conjur.org/Latest/en/Content/ReleaseNotes/ConjurOSS-suite-RN.htm

Reviewer 2: "+" -> and later

> - Conjur Enterprise v5.6.3+

Reviewer 2: ```suggestion
- Conjur Enterprise v11.2.1 (v5.6.3) and later
```

### Sypported Platforms

- macOS Catalina+
- Windows 10+
- Red Hat Enterprise Linux 7/8

### Installation

#### Install the CLI

You can access the latest release of the CLI to install on your machine by navigating to our 
[release](https://github.com/cyberark/conjur-api-python3/releases) page. For instructions on 

> how to setup and configure the CLI, see our official documentation ***(TODO add link to docs)*** 

Reviewer 2: Conjur CLI

Reviewer 2: link:
https://docs.conjur.org/Latest/en/Content/Developer/CLI/cli-lp.htm

Author: I get "The resource you are looking for has been removed, had its name changed, or is temporarily unavailable."

This will be live when u push?

#### Install the SDK

The SDK can be installed via PyPI. Note that the SDK is a **Community** level project meaning that the SDK is subject 
to alterations that may result in breaking change. To avoid breaking changes, make sure that you stay up-to-date on our 

> latest releases and review the project's CHANGELOG.md](CHANGELOG.md).

Reviewer 3: ```suggestion
latest releases and review the project's [CHANGELOG.md](CHANGELOG.md).
```

> ### CLI

Reviewer 2: ```suggestion
### Conjur CLI
```

> For more information on how to setup, configure, and start using the CLI, see our official 

Reviewer 2: ```suggestion
For more information on how to setup, configure, and start using the Conjur CLI, see our official 
```

> documentation found here ***(TODO add link to docs)***.

Reviewer 2: https://docs.conjur.org/Latest/en/Content/Developer/CLI/cli-lp.htm

                ca_bundle='/path/to/my/ca/bundle.pem')
```

#### With `.netrc` and `.conjurrc` files

You can provide the above parameters passed in as part of the Client initialization process in two files 
(`.netrc` and `.conjurrc`). Both `.netrc` and `.conjurrc` files should be saved to your home directory. 
For example `~/.netrc`. 

The `.conjurrc` contains connection details needed to connect to the Conjur endpoint and should 
contain details for cert_file, conjur_account, and conjur_url.

```
# .conjurrc
---
cert_file: /Users/someuser/conjur-server.pem
conjur_account: someaccount
conjur_url: https://conjur-server
```

The `.netrc` file or (`_netrc` for Windows environments) contains credentials needed to login the 
Conjur endpoint and should consist of machine, login, and password. Note that the value for 
'machine' should have `/authn` appended to the end of the url and 'password' should contain the 

> API key of the user you intend to log in as.

Reviewer: @sigalsax I think this should be reviewed also by @sashaCher or @moticless

Author: @InbalZilberman what do we need them to review here? writing this or providing this option?

Author: @InbalZilberman in terms of what do we want to check?
@moticless and I spoke about this and we dont have a problem with writing this as long as we tell the user to follow least privilege and ensures that no other user can access this file

print("Variable value is:", new_value.decode('utf-8'))
```

#### With login ID and API key

> Write the code same as in the first example but initialize the Client with the following arguments:

Reviewer 2: that is so american :)

I would say 
Write the code as in the first example
or 
Write the code in the same way as in the first example

but the american was is fine too, i think

                ca_bundle='/path/to/my/ca/bundle.pem')
```

#### `.conjurrc` file

You can provide the connection details needed to connect to the Conjur endpoint (`url`, `account`, and `ca_bundle`) in 
a file called `.conjurrc` instead of passing them in during initialization of the Client.

The `.conjurrc` file should be saved to your home directory and should contain `conjur_url`, 
`conjur_account`, and`cert_file`.

```
# .conjurrc
---
cert_file: /Users/someuser/conjur-server.pem
conjur_account: someaccount
conjur_url: https://conjur-server
```

#### Storing credentials

When using the SDK, you can keep credentials in the system's native credential store instead of passing them in 
during initialization of the Client. By default, the Client will favor saving credentials (login ID and password) 
to the credential store. If the detected credential store is not one we support or is not accessible, the 
credentials will be written to a config file, netrc, in plaintext. 

> If written in the netrc it is highly recommended that you remove the file when done.

Reviewer 2: ```suggestion
If written in the netrc, it is strongly recommended that you log out when not using the Conjur CLI. This removes the credentials from the local cache.

```

Author: updated to "This removes the credentials from the `.netrc` file." because I think we should refrain from saying using "local cache"

A configuration file called `.conjurrc` is used to hold details required to communicate 
to the Conjur server. You can provide these details needed to open a connection to the 
Conjur endpoint in this file instead of passing them in (`url`, `account`, and `ca_bundle`)  
during initialization of the Client.

The `.conjurrc` file should be saved to your home directory and should contain `conjur_url`, 
`conjur_account`, and`cert_file`.

```
# .conjurrc
---
cert_file: /Users/someuser/conjur-server.pem
conjur_account: someaccount
conjur_url: https://conjur-server
```

#### Storing credentials

When using the SDK, you can keep credentials in the system's native credential store 
instead of passing them in during initialization of the Client. By default, the Client 
will favor saving credentials (login ID and password) to the system's credential store. 
If the detected credential store is not one we support or is not accessible, the 
credentials will be written to a configuration file, `.netrc`, in plaintext. 

If written to the `.netrc`, it is strongly recommended that you log out when not 
using the Conjur CLI. This removes the credentials from the `.netrc` file.

> The `.netrc` file or (`_netrc` for Windows environments) contains credentials needed to login the 

Reviewer 2: ```suggestion
The `.netrc` file or (`_netrc` for Windows environments) contains credentials needed to log in the 
```

> Conjur endpoint and should consist of machine, login, and password.

Reviewer 2: ```suggestion
Conjur endpoint and should consist of 'machine', 'login', and 'password'.
```

> Note: Policy to create the variable must have been already loaded

Reviewer 2: ```suggestion
Note: Policy to create the variable must have already been loaded
```

otherwise you will get a 404 error during invocation.

#### `load_policy_file(policy_name, policy_file)`

Applies a file-based YAML to a named policy. This method only

> supports additive changes. Result is a dictionary object constructed

Reviewer 2: This method supports additive changes only.

Author: this is a name of a function :)

> usually a destructive invocation. Result is a dictionary object

Reviewer 2: Replaces a named policy with a policy from the provided file.

Author: what do you mean? I think it is self explanation when we say destructive

| List constraints | Explanation                                                  |
| ---------------- | ------------------------------------------------------------ |
| kind             | Filter resources by specified kind (user, host, layer, group, policy, variable, or webservice) |
| limit            | Limit list of resources to specified number                  |
| offset           | Skip specified number of resources                           |
| role             | Retrieve list of resources that specified role is entitled to see (must specify role’s full ID) |
| search           | Search for resources based on specified query                |
| inspect          | List the metadata for resources                              |

#### `rotate_other_api_key(resource: Resource)`

Rotates another entity's API key and returns it as a string.

Note: resource is of type Resource which should have `type` (user / host) and 
`name` attributes.

#### `rotate_personal_api_key(logged_in_user, current_password)`

Rotates the personal API key of the logged-in user and returns it as a string.

#### `change_personal_password(logged_in_user, current_password, new_password)`

> Updates the current, the logged in user's password with the password parameter provided. 

Reviewer 2: ```suggestion
Updates the current, logged-in user's password with the password parameter provided. 
```

Note: the new password must meet the Conjur password complexity constraints. It must 
contain at least 12 characters: 2 uppercase, 2 lowercase, 1 digit, 1 special character.

#### `whoami()`

> _Note: This method requires Conjur v1.9+_

Reviewer 2: @sigalsax  - I don;t see a conjur oss version 1.9. Please see list of releases and consult with @izgeri

Reviewer 2: @sigalsax  - what about this?
@izgeri ??

Author: it is 1.9
<img width="910" alt="Screen Shot 2021-04-12 at 9 14 32 AM" src="https://user-images.githubusercontent.com/19418506/114348830-c0028d80-9b6f-11eb-979e-707831271ceb.png">

Author: @InbalZilberman

usually a destructive invocation. Result is a dictionary object
constructed from the returned JSON data.

#### `update_policy_file(policy_name, policy_file)`

Modifies an existing Conjur policy. Data may be explicitly
deleted using the `!delete`, `!revoke`, and `!deny` statements. Unlike
"replace" mode, no data is ever implicitly deleted. Result is a
dictionary object constructed from the returned JSON data.

#### `list(list_constraints)`

Returns a list of all available resources for the current
account.

> The list constraints parameter is optional and should be provided as

Reviewer 2: ```suggestion
The 'list constraints' parameter is optional and should be provided as
```

> API request (such as its ip address, user, account,

Reviewer 2: client has been with a capital C all through this docs until this instance. I think this should be Client too.

> API request (such as its ip address, user, account,

Reviewer 2: ```suggestion
API request (such as its IP address, user, account,
```

> token expiration date etc.).

Reviewer 2: ```suggestion
token expiration date, etc).
```
